backend/canisters/user_index/impl/src/updates/c2c_suspend_users.rs

```diff
@@ -0,0 +1,45 @@
+use crate::guards::caller_is_group_index;
+use crate::model::set_user_suspended_queue::SetUserSuspended;
+use crate::{mutate_state, RuntimeState};
+use canister_api_macros::update_msgpack;
+use canister_tracing_macros::trace;
+use user_index_canister::c2c_suspend_users::{Response::*, *};
+
+#[update_msgpack(guard = "caller_is_group_index")]
+#[trace]
+fn c2c_suspend_users(args: Args) -> Response {
+    mutate_state(|state| c2c_suspend_users_impl(args, state))
+}
+
+fn c2c_suspend_users_impl(args: Args, runtime_state: &mut RuntimeState) -> Response {
+    let now = runtime_state.env.now();
+
+    let suspended_until = args.duration.map(|d| now + d);
+
+    let users_to_suspend: Vec<_> = args
+        .user_ids
+        .into_iter()
+        .filter(|u| {
+            runtime_state
+                .data
+                .users
+                .suspend_user(u, suspended_until, args.reason.clone(), now)
+        })
+        .collect();
+
+    runtime_state.data.set_user_suspended_queue.enqueue(
+        users_to_suspend
+            .iter()
+            .map(|u| SetUserSuspended::User(*u, suspended_until, args.reason.clone()))
+            .collect(),
+    );
+
+    if let Some(ts) = suspended_until {
+        runtime_state
+            .data
+            .set_user_suspended_queue
+            .schedule(users_to_suspend.iter().map(|u| SetUserSuspended::Unsuspend(*u)).collect(), ts);
+    }
+
+    Success
+}
```

Comment:
Presumably we also want to queue up a UserEvent for each suspended user like we do in suspend_user:

runtime_state.data.user_event_sync_queue.push(
        user_id,
        UserEvent::UserSuspended(UserSuspended {
            timestamp: now,
            duration: duration.map_or(SuspensionDuration::Indefinitely, SuspensionDuration::Duration),
            reason,
        }),
    );

Reply:
Here we're just queuing them all up to be handled later.
They then get processed during heartbeat and we simply call into `suspend_user`.
See here - https://github.com/open-ic/open-chat/blob/d60744087a8077115abc321677d8da7f5798befa/backend/canisters/user_index/impl/src/lifecycle/heartbeat.rs#L425

Reply:
Doh! I've removed that duplication now.